pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -54,230 +56,170 @@ func (e transientError) Error() string {
-	}
-	// check if the latest deployment exists
-	// we'll return after we've dealt with the multiple-active-deployments case
-	latestDeploymentExists, latestDeploymentStatus := deployutil.LatestDeploymentInfo(config, existingDeployments)
-	if latestDeploymentExists && latestDeploymentStatus != deployapi.DeploymentStatusFailed {
-		// If the latest deployment exists and is not failed, clear the dc message
-		return "", existingDeployments, latestDeploymentExists, nil
-	}
-
-	// Rest of the code will handle non-existing or failed latest deployment causes
-	// TODO: Inspect pod logs in case of failed latest
-
-	details := ""
-	allDetails := []string{}
-	if config.Details != nil && len(config.Details.Message) > 0 {
-		// Populate details with the previous message so that in case we stumble upon
-		// an unexpected client error, the message won't be overwritten
-		details = config.Details.Message
-	}
-	// Look into image change triggers and find out possible deployment failures such as
-	// missing image stream tags with or without build configurations pointing at them
-	for _, trigger := range config.Triggers {
-		if trigger.Type != deployapi.DeploymentTriggerOnImageChange || trigger.ImageChangeParams == nil || !trigger.ImageChangeParams.Automatic {
-			continue
-		}
-		name := trigger.ImageChangeParams.From.Name
-		tag := trigger.ImageChangeParams.Tag
-		istag := imageapi.JoinImageStreamTag(name, tag)
-
-		// Check if the image stream tag pointed by the trigger exists
-		if _, err := c.osClient.ImageStreamTags(config.Namespace).Get(name, tag); err != nil {
-			if !errors.IsNotFound(err) {
-				glog.V(2).Infof("Error while trying to get image stream tag %q: %v", istag, err)
-				return details, existingDeployments, latestDeploymentExists, nil
+// reconcileDeployments reconciles existing deployment replica counts which
+// could have diverged outside the deployment process (e.g. due to auto or
+// manual scaling, or partial deployments). The active deployment is the last
+// successful deployment, not necessarily the latest in terms of the config
+// version. The active deployment replica count should follow the config, and
+// all other deployments should be scaled to zero.
+//
+// Previously, scaling behavior was that the config replica count was used
+// only for initial deployments and the active deployment had to be scaled up
+// directly. To continue supporting that old behavior we must detect when the
+// deployment has been directly manipulated, and if so, preserve the directly
+// updated value and sync the config with the deployment.
+func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *kapi.ReplicationControllerList, config *deployapi.DeploymentConfig) error {
+	activeDeployment := deployutil.ActiveDeployment(config, existingDeployments)
+	_, latestDeployment := deployutil.LatestDeploymentInfo(config, existingDeployments)
+	activeDeploymentIsLatest := activeDeployment.Name == latestDeployment.Name
```

Comment:
I'll clean this up to assert some preconditions; should also further simplify the code